cloudevents/pydantic/v2/event.py

```diff
@@ -0,0 +1,222 @@
+
+            If no attribute is given the class MUST use the kwargs as the attributes.
+
+            Example Attributes:
+            {
+                "specversion": "1.0",
+                "type": "com.github.pull_request.opened",
+                "source": "https://github.com/cloudevents/spec/pull",
+                "id": "A234-1234-1234",
+                "time": "2018-04-05T17:31:00Z",
+            }
+
+        :param data: Domain-specific information about the occurrence.
+        """
+        if attributes:
+            if len(kwargs) != 0:
+                # To prevent API complexity and confusion.
+                raise IncompatibleArgumentsError(
+                    "Attributes dict and kwargs are incompatible."
+                )
+            attributes = {k.lower(): v for k, v in attributes.items()}
+            kwargs.update(attributes)
+        super(CloudEvent, self).__init__(data=data, **kwargs)
+
+    model_config = ConfigDict(
+        extra="allow",  # this is the way we implement extensions
+        json_schema_extra={
+            "example": {
+                "specversion": "1.0",
+                "type": "com.github.pull_request.opened",
+                "source": "https://github.com/cloudevents/spec/pull",
+                "subject": "123",
+                "id": "A234-1234-1234",
+                "time": "2018-04-05T17:31:00Z",
+                "comexampleextension1": "value",
+                "comexampleothervalue": 5,
+                "datacontenttype": "text/xml",
+                "data": '<much wow="xml"/>',
+            }
+        },
+    )
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_base64_data_input(cls, data: typing.Any) -> typing.Any:
+        """Populates the `data` property if the model gets created using `data_base64`.
+
+        :param data: Input data.
+
+        :return: Event serialized as a standard CloudEvent dict with user specific
+        parameters.
+        """
+        if isinstance(data, dict) and data.get("data_base64") is not None:
+            data["data"] = base64.b64decode(data["data_base64"])
+            del data["data_base64"]
+        return data
+
+    @model_serializer(when_used="json", mode="wrap")
+    def serialize_model(self, handler: typing.Callable) -> typing.Dict[str, typing.Any]:
+        """Performs Pydantic-specific serialization of the event.
+
+        Needed by the pydantic base-model to serialize the event correctly to json.
+        Without this function the data will be incorrectly serialized.
+
+        :param self: CloudEvent.
+        :param handler: The nested serialization handler.
+
+        :return: Event serialized as a standard CloudEvent dict with user specific
+        parameters.
+        """
+        model_dump: dict = handler(self)
+        if isinstance(self.data, (bytes, bytearray, memoryview)):
+            model_dump["data_base64"] = base64.b64encode(self.data).decode("ascii")
+        return model_dump
```

Comment:
If data is a nested BaseModel, this logic could produce a non compliant JSON string.

Do we assume that when serializing to JSON the contentdatatype is `application/json`? Then `data` should be serialized in a JSON string before serializing the model.

The concept of assumption sounds wrong, though. Perhaps I can add a `serialize_data` method that can be overridden, or, even better a small dictionary that acts as a registry (so people can create other base classes supporting multiple contentdatatypes.

```python
def json_data_serializer(data):
    return json_serialized_data
    
class CloudEvent(abstract.CloudEvent, BaseModel):
    serializer_registry = {
        "application/json": json_data_serializer
    }
    
    ...
```

Reply:
I answered this in the previous comment. I'll address this function as well.